frontend/server/k8s-helper.ts

```diff
@@ -127,3 +132,79 @@ export function getPodLogs(podName: string): Promise<string> {
       (error: any) => { throw new Error(JSON.stringify(error.body)); }
     );
 }
+
+/**
+* Create Tensorboard pod via CRD with the given logdir if there is no existing
+* Tensorboard pod.
+*/
+export async function newTensorboardInstance(logdir: string): Promise<void> {
+  if (!k8sV1CustomObjectClient) {
+    throw new Error('Cannot access kubernetes Custom Object API');
+  }
+  const currentPod = await getTensorboardInstance(logdir);
+  if (currentPod) {
+    return;
+  }
+
+  // TODO: take the configuration below to a separate file
+  // Name of the viewer resource is based on logDir.
+  const group = 'kubeflow.org';
+  const version = 'v1beta1';
+  const plural = 'viewers';
+  const body = {
+    apiVersion: group + '/' + version,
+    kind: 'Viewer',
+    metadata: {
+      name: 'viewer-' + crypto.SHA1(logdir),
+      namespace: namespace,
+    },
+    spec: {
+      type: 'tensorboard',
+      tensorboardSpec: {
+        logDir: logdir,
+      }
+    }
+  };
+  await k8sV1CustomObjectClient.createNamespacedCustomObject(group, version,
+    namespace, plural, body);
+}
+
+/**
+* Finds a running Tensorboard pod created via CRD with the give logdir and
+* returns its pod IP and port.
+*/
+export async function getTensorboardInstance(logdir: string): Promise<string> {
+  if (!k8sV1CustomObjectClient) {
+    throw new Error('Cannot access kubernetes Custom Object API');
+  }
+
+  const group = 'kubeflow.org';
+  const version = 'v1beta1';
+  const plural = 'viewers';
+  const pods = (await k8sV1CustomObjectClient.listNamespacedCustomObject(group,
+    version, namespace, plural)).body.items;
+  const pod = pods.find((p) =>
+    p.metadata.name == 'viewer-' + crypto.SHA1(logdir) &&
```

Comment:
will crypto.SHA1(logdir) have size limit?
the metadata name have size limit of 256 as far as i know. it might fail if the length of the logdir is too long

Reply:
My understanding is that the output length of sha1 is fixed at 160 bit as long as the input message is < 2^64 bits. https://nvlpubs.nist.gov/nistpubs/FIPS/NIST.FIPS.180-4.pdf